django-cms/djangocms-bootstrap4
# Quick: Fix Grid Container get_short_description for Nested Tuples Setting

## To Do

<details><summary>Done</summary>

- [x] (As necessary) Add test cases. ([direction](https://github.com/django-cms/djangocms-bootstrap4/pull/145#issuecomment-1147687443)) ([done by mark](https://github.com/django-cms/djangocms-bootstrap4/pull/145#issuecomment-1148125883) ([merged?](https://github.com/django-cms/djangocms-bootstrap4/pull/145/commits/58372d7d719356a9911db46e21710dd8c2f3997c))
- [x] The `get_short_description` value is inaccurate (see [screenshot](https://user-images.githubusercontent.com/62723358/172020564-e71e542e-1164-42fe-a450-18838bc99c37.png)). (4ba6861, 99dcc07).
- [x] Test changes from review (da1f543, 8d8adcb, 4ba6861).

</details>

## Description

Fix short description for Bootstrap4 Container if user choses container type from nested tuples in `DJANGOCMS_BOOTSTRAP4_GRID_CONTAINERS` setting.\*

_\* An `<optgroup>` is available in "Container Type" `<select>` field when `DJANGOCMS_BOOTSTRAP4_GRID_CONTAINERS` has deeply-nested tuples. But were user to chose such an option, then `get_short_descritpion` would return empty string._

## Related resources

* [DJANGO: ModelChoiceField optgroup tag | StackOverflow](https://stackoverflow.com/a/17854288/11817077)
* [discussion about testing backward compatibility with existing install of plugin](https://django-cmsworkspace.slack.com/archives/C01J7H87KFC/p1642531751019400)

## Checklist

* [x] I have opened this pull request against ``master``
    <details>My modification is indirect. I changed default value for a constant (settings fallback value), which affects one test.</details>
* [x] I have read the [contribution guidelines ](https://github.com/django-cms/django-cms/blob/develop/CONTRIBUTING.rst) and I have joined #workgroup-pr-review on 
[Slack](https://www.django-cms.org/slack) to find a “pr review buddy” who is going to review my pull request.
    <details><a href="https://django-cmsworkspace.slack.com/archives/C01J7H87KFC/p1642523992015800" target=_blank">my request for review</a></details>

## Testing

1. Setup a Django CMS and this branch of djangocms-bootstrap4, e.g.
    - install [**wesleyboar/djangocms-bootstrap4** at `wesleyboar:quick/grid-container-support-nested-tuple-choices`](https://github.com/wesleyboar/djangocms-bootstrap4/tree/quick/grid-container-support-nested-tuple-choices)
    - or use [**wesleyboar/django-cms-quickstart** at `poc/djangocms-bootstrap4-pull-145`](https://github.com/wesleyboar/django-cms-quickstart/tree/poc/djangocms-bootstrap4-pull-145#django-cms-quickstart) (see [diff](https://github.com/wesleyboar/django-cms-quickstart/compare/main...wesleyboar:poc/djangocms-bootstrap4-pull-145)).[^1][^2]
2. Create a page in the CMS.
5. Create Container plugin instances on the page for each CONTAINER TYPEs.
6. **✓ Verify** the class in the markup of each element added.
7. **⊖ Verify** the short description (parenthesized value in Structure mode) is accurate.

| choice | class | description |
| - | - | - |
| Default → Container | `container` | (Container) |
| Default → Fluid container | `container-fluid` | (Fluid container) |
| Custom → Your container | `container-yours` | (Your container) |

## Screenshots

![Bootstrap Grid Container Nested Tuple](https://user-images.githubusercontent.com/62723358/172215747-07a326bd-2794-49d6-b1c9-0dbc69d13eca.png)


[^1]: Added `from django.utils.translation import gettext_lazy as _` to `your-app/settings.py`.
[^2]: Added [sample DJANGOCMS_BOOTSTRAP4_GRID_CONTAINERS](https://github.com/django-cms/djangocms-bootstrap4/pull/145/files#diff-7b3ed02bc73dc06b7db906cf97aa91dec2b2eb21f2d92bc5caa761df5bbc168fR127-R133) to `your-app/settings.py`.
